cli/tests/semgrep_runner.py

```diff
@@ -0,0 +1,113 @@
+#
+# Run either semgrep in the same process using Click or inovke the
+# mlgrep command, which reimplements the semgrep CLI in OCaml.
+#
+# This is replacement for CliRunner provided by Click.
+#
+import os
+import sys
+from subprocess import PIPE
+from subprocess import Popen
+from typing import Dict
+from typing import List
+from typing import Optional
+from typing import Sequence
+from typing import Union
+
+from click.testing import CliRunner
+
+# Environment variables that trigger the use of mlgrep
+MLGREP_PATH = "mlgrep"
+env_mlgrep = os.environ.get("PYTEST_MLGREP")
+if env_mlgrep:
+    MLGREP_PATH = env_mlgrep
+USE_MLGREP = True if os.environ.get("PYTEST_USE_MLGREP") else False
+
+# The semgrep command suitable to run semgrep as a separate process.
+# It's something like ["semgrep"] or ["python3"; -m; "semgrep"] or
+# ["/path/to/mlgrep"].
+SEMGREP_BASE_COMMAND: List[str] = (
+    [MLGREP_PATH] if USE_MLGREP else [sys.executable, "-m", "semgrep"]
+)
+
+SEMGREP_BASE_COMMAND_STR: str = " ".join(SEMGREP_BASE_COMMAND)
+
+
+class Result:
+    """Minimal properties of click.testing.Result used in our project.
+
+    https://click.palletsprojects.com/en/8.0.x/api/#click.testing.Result
+    """
+
+    def __init__(self, exit_code: int, stdout: str, stderr: str):
+        self._exit_code = exit_code
+        self._stdout = stdout
+        self._stderr = stderr
+
+    @property
+    def stdout(self) -> str:
+        return self._stdout
+
+    @property
+    def stderr(self) -> str:
+        return self._stderr
+
+    @property
+    def output(self) -> str:
+        """alias for stdout"""
+        return self._stdout
+
+    @property
+    def exit_code(self) -> int:
+        return self._exit_code
+
+
+# Run mlgrep in an external process
+# (there's no other choice since it's an OCaml binary)
+def invoke_mlgrep(
+    args: Optional[Union[str, Sequence[str]]], env: Optional[Dict[str, str]] = None
+) -> Result:
+    arg_list: List[str] = []
+    if isinstance(args, str):
+        # This is not exactly how the shell would interpret the arguments
+        # but hopefully it's sufficient to run our tests correctly.
+        # If this doesn't work, consider changing the tests to they pass
+        # arguments as a list rather than a string.
+        arg_list = args.split(" ")
+    elif isinstance(args, List):
+        arg_list = args
+    argv: List[str] = [MLGREP_PATH] + arg_list
+    env_dict = {}
+    if env:
+        env_dict = env
+    full_env = dict(os.environ, **env_dict)
+    proc = Popen(argv, stdout=PIPE, stderr=PIPE, env=full_env)
+    stdout, stderr = proc.communicate()
+    return Result(proc.returncode, stdout.decode("utf-8"), stderr.decode("utf-8"))
+
+
+class SemgrepRunner:
+    """Run either semgrep with CliRunner or with mlgrep.
+
+    It's meant as a drop-in replacement for CliRunner(cli, args).
+    If a property is missing on the runner object, please add it here.
```

Comment:
what do you mean by property of runner object? Do you have an example?

Reply:
The runner object is the one created by `CliRunner(cli, args)`, obviously. You need to understand what `CliRunner` does. I explained it briefly in the `Result` class above. It's a hack that pretends to start a new Python program but doesn't, saving the startup time. Not my creation. See https://click.palletsprojects.com/en/8.1.x/testing/